src/psyclone/psyir/nodes/operation.py

```diff
@@ -224,38 +230,137 @@ def create(operator, lhs, rhs):
         child instances with optional names.
 
         :param operator: the operator used in the operation.
-        :type operator: \
+        :type operator:
             :py:class:`psyclone.psyir.nodes.BinaryOperation.Operator`
-        :param lhs: the PSyIR node containing the left hand side of \
-            the assignment, or a tuple containing the name of the \
+        :param lhs: the PSyIR node containing the left hand side of
+            the assignment, or a tuple containing the name of the
             argument and the PSyIR node.
-        :type lhs: Union[:py:class:`psyclone.psyir.nodes.Node`, \
+        :type lhs: Union[:py:class:`psyclone.psyir.nodes.Node`,
             Tuple[str, :py:class:`psyclone.psyir.nodes.Node`]]
-        :param rhs: the PSyIR node containing the right hand side of \
-            the assignment, or a tuple containing the name of the \
+        :param rhs: the PSyIR node containing the right hand side of
+            the assignment, or a tuple containing the name of the
             argument and the PSyIR node.
-        :type rhs: Union[:py:class:`psyclone.psyir.nodes.Node`, \
+        :type rhs: Union[:py:class:`psyclone.psyir.nodes.Node`,
             Tuple[str, :py:class:`psyclone.psyir.nodes.Node`]]
 
         :returns: a BinaryOperator instance.
         :rtype: :py:class:`psyclone.psyir.nodes.BinaryOperation`
 
-        :raises GenerationError: if the arguments to the create method \
+        :raises GenerationError: if the arguments to the create method
             are not of the expected type.
 
         '''
-        if not isinstance(operator, Enum) or \
-           operator not in BinaryOperation.Operator:
+        if (not isinstance(operator, Enum) or
+                operator not in BinaryOperation.Operator):
             raise GenerationError(
                 f"operator argument in create method of BinaryOperation class "
                 f"should be a PSyIR BinaryOperation Operator but found "
                 f"'{type(operator).__name__}'.")
-
         binary_op = BinaryOperation(operator)
         binary_op.addchild(lhs)
         binary_op.addchild(rhs)
         return binary_op
 
+    @property
+    def datatype(self):
+        '''
+        :returns: the datatype of the result of this BinaryOperation.
+        :rtype: :py:class:`psyclone.psyir.symbols.DataType`
+
+        :raises InternalError: if the types of the operands are not permitted
+                               for the given operator.
+        :raises NotImplementedError: if the operands are of the same intrinsic
+                                     type but different precisions.
+        :raises TypeError: if the operands are both arrays but are of different
+                           shapes.
```

Comment:
Again, instead of `UnknownFortranType("")` consider `DeferredType()`, I am not so sure about the errors, returning Deferred will make things easier by removing special casing of the caller and better chaining, but we will lose the information about why it happens (we don't have logging yet :/)

Reply:
The UnknownFortranTypes are now all DeferredType. I think we need the `InternalError` and `TypeError` (could also be InternalError?) because, in those situations, we are confident that something has gone wrong. I take your point about the `NotImplementedError` though and will change that.